BOOTSEQ.md

```diff
@@ -4,77 +4,160 @@
-- There are use cases for network boot only.
+- Resetting the xPU is assumed to cause PCIe errors that may impact the operation of the host OS.
+  - There are different kinds of resets. Generally XPU core complex reset and XPU PCIe MAC reset should be separated for minimal disruptive ISSU etc. but a complete SoC reset of the xPU will cause PCIe errors and may result in a controlled or uncontrolled host reset.
+  - See  _Host Reset or Crash_ and _xPU Reset or Crash_ sections for specifics.
+
+- There are use cases for network boot only where the host is used only to provide power and cooling to one or more xPUs.
 - xPU boot can take some time (full linux distro)
 - There is communication between some of the entities from above
 - More...
 
-## Bootup
+## Simultaneous Host and xPU Bootup
+
+1. Host and xPU are simultaneously powered up
+2. Host and xPU are simultaneously restarted
+
+How tightly coupled should host and xPU be during boot phase?
+
+- This is likely use-case dependent, so OPI should accomodate all cases from tightly coupled to completely independent operation.
+
+Events:-
 
-- Server is Powered On
+- Host is Powered On
 - xPU receives power and starts booting
-  - These events may not be synchronous. The xPU may be installed in a PCIe slot providing persistent power. xPU power cycle may be a function of the host BMC (and potentially the xPU BMC) rather than the power state of the host
-- Should Host OS wait for xPU to finish boot ?
+- Should host OS wait for xPU to finish boot ?
   - For some functions, it must (e.g. NVMe namespace that it needs to boot from)
   - With a pre-provisioned, local host OS present, and for other functions (most notably, networking), the host can continue its boot sequence. Ports should appear down from the host's perspective until the xPU is ready to handle traffic from/to host.
-  - Host OS continues to its boot. xPU needs to either respond to PCIe transactions or respond with retry config cycles(CRS) till it is ready. This automatically holds the Host BIOS. Note that Host BIOS may need to be modified to allow longer CRS response time before timing out.
-- Should xPU OS wait for host to finish boot ?
-  - How tightly coupled should host and xPU be during boot phase?
-- who tells Host OS/xPU OS to wait and how ?
-  - PCIe has a mechanism for config retry cycles(CRS) that can be used.
-- who tells Host OS to to continue / start booting ?
-  - These events may not be synchronous. The xPU may be installed in a PICe slot providing persistent power. xPU power cycle may be a function of the host BMC (and potentially the xPU BMC) rather than the power state of the host
-- is Host CPU halted ?
-- can we wait in UEFI/BIOS in case of network boot via xPU ?
+  - Host OS continues to its boot. xPU needs to either respond to PCIe transactions or respond with retry config cycles (CRS) till it is ready. This automatically holds the host BIOS.
```

Comment:
Agreed.  CRS is not practical.  Some xPUs may take as long as 5 mins after power is applied to respond to PCIe Config Cycles.  The x86 CPU will timeout and crash in 10s of milliseconds.  There is a another PCIe retry mechanism when the target interrupts the x86 host when it is ready, but BIOS is not ready to accept interrupts that earily in the boot process.

Reply:
@RezaBacchus unless we change the BIOS to longer timeouts, this is what we did at DELL... will that work ?

Reply:
In some implementations, while the xPU is initializing the PCIe manager, the xPU ignores config cycles leading BIOS to infer that the card is absent.  In other cases, the xPU claims the config cycle but fails to respond causing a catastrophic reset due to a Table of Record (ToR) timeout in the x86 CPU.  CRS assumes that the PCIe manager on the xPU is stable enough to avoid these 2 cases.  
The same behavior is observed when the ARM cores lock up during runtime. A PF0 HW PCIe register that is independent of SW running on the ARM is a solution to both the boot-up and runtime issues.  This however will require the PCI SIG to standardize on the appropriate capabilities and status registers; and will take time to implement in silicon.

Reply:
@RezaBacchus This would assume that PCIe subsystem is independent from Arm complex - at least to the point where it can setup config space on its own. This is not the case for Intel IPU, as IMC configures PCIe subsystem (including config space), including releasing PCIe PHY from reset and enabling link training SSM.

Considering how implementations of PCIe subsystem may vary across xPU vendors, including CRS reliability concerns, CRS might not be the way to go.

Reply:
good with options that does any coordination with BIOS/BMC - going in interdependent lock-step will ensure the dependencies are cleared.